### DISCUSSION TITLE
Login with Google Provider working in Firefox but failing in Chrome/Safari

### DISCUSSION THREAD

**Discussion by seiwaldm** [pocketbase/pocketbase#966] [Q&A] [answered]
Hi! 

I'm using Pocketbase as a Backend Solution for a little Card Game based on data from Superhero API. The source Code can be found [here](https://github.com/seiwaldm/heroclash-vitesse), the deployed version is [hosted on Netlify](https://heroclash.seiwald.club/). I managed to set up Login/Registration via OAuth-Providers (Google, GitHub). Both providers work fine in Firefox. 

However, when I try to sign in via Google in Chrome, I get a "ClientResponseError 400: Failed to authenticate." Response without any further information. Seemst to be the same Problem in Safari (and Edge with its Chromium Base), although I don't have personal access to an Apple device. Anyone can point me to the right direction? What am I doing wrong? The error should be easy to replicate. 

Thanks in advance for your efforts

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure what could be reason for the error, but you could try starting the PocketBase executable with `--debug` and/or check the request logs from "Admin UI > Logs" for a more a detailed error message and post it here.

**Reply by ganigeorgiev** [reply]
Alternatively, to rule out some issue with your integration, you could try replicating the example from https://pocketbase.io/docs/manage-users/#auth-via-oauth2:
1. Create 2 files `pb_public/index.html` and `pb_public/redirect.html` and copy the content from the example (for the sdk script url you can use https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.umd.js).
2. Update your redirect url in the Google and GitHub OAuth2 app settings

**Reply by seiwaldm** [reply]
Thank you very much @ganigeorgiev! Of course it was a client side mistake and you pointed me into the right direction. Checking the Admin UI logs gave a hint to me. Refactored my code and now I got a cleaner design and no more async-problems... just donated for the project! Great work and thx for the quick support!

**Reply by hzhu** [reply]
I'm running into the same issue as you described. I was curious what specifically was the issue? What hint did you see? 

<img width="692" alt="Screenshot 2023-10-09 at 22 58 15" src="https://github.com/pocketbase/pocketbase/assets/1811365/b7eb2e6b-316c-44a0-87c0-a71a1fd9e5ed">


<img width="691" alt="Screenshot 2023-10-09 at 22 20 59" src="https://github.com/pocketbase/pocketbase/assets/147226466/c27273c1-bea8-4e69-bf98-ae2ced890552">

cc. @seiwaldm @ganigeorgiev 🙏

**Reply by ganigeorgiev** [reply]
@hzhu I'm not sure how to help you without a minimal code sample of your implementation or at least some information how did you setup the Google OAuth2 app.

From the above screenshots and the `node-fetch` user agents I guess you are using the "Manual code exchange" flow, right? In this case double check that you are extracting and sending the correct parameters to the `authWithOAuth2Code` call in your node redirect handler.

**Reply by seiwaldm** [reply]
Hey there @hzhu! As mentioned in my original comment, I made a mistake on the client. In my first version I tried to redirect to the Google Login via the href-Attribute of an html ``<a>``-Element, where the URL for the attribute was dynamically injected after a call to the asynchronous ``listAuthMethods()``-Method provided by Pocketbase. 

This flawed approach did actually work in Firefox, because the Browser was somehow slow enough that the async method call returned before the browser executed the navigation to the URL. However, Chrome was too fast for this (at least that was the explanation I gave to myself...).

So I fixed the problem by replacing the ``<a>``-Element with a ``<button>``, explicitly waiting for the ``listAuthMethods()``-Method to return and only then navigating to the Google Login via Javascript.

![grafik](https://github.com/pocketbase/pocketbase/assets/15251387/045a3fc3-9457-4adb-8a7d-a25b638b8352)
![grafik](https://github.com/pocketbase/pocketbase/assets/15251387/591ba8e0-5b3b-4ab2-a52c-10e4dae6e1d9)

Good luck with fixing your problem!

**Reply by hzhu** [reply]
Thank you for the explanation, @seiwaldm! While I haven't pinpointed the exact cause, it seems to be a related issue.

Hello @ganigeorgiev! I've set up a minimal [code example](https://github.com/hzhu/remix-pocketbase-oauth). This project was bootstrapped using the [create-remix](https://remix.run/docs/en/main/other-api/create-remix) app (I'm trying to bring PocketBase to Remix!) and integrates with PocketBase. Additionally, you can check out this app in its 🌐 production environment here: https://remix-pocketbase-oauth.vercel.app. The associated database is hosted at [hop.io](https://hop.io/).

The quick start steps can be found in the [README.md](https://github.com/hzhu/remix-pocketbase-oauth/tree/main#readme).

## Steps to Reproduce the Bug
1. Visit the production app (or visit it at http://localhost:3000 if ran locally).
1. Select "Login", followed by "Continue with Google".
1. Provide your Google account credentials.
1. Complete the login sequence.
1. Occasionally, you will find that the login fails.


If the bug is triggered, an error message stating `Failed to authenticate.` will appear in the UI. This issue arises [here](https://github.com/hzhu/remix-pocketbase-oauth/blob/main/app/routes/auth.google.callback.tsx#L52), during the `.authWithOAuth2Code` call. The admin UI shows this error:

```
{
  "errorDetails": "constraint failed: UNIQUE constraint failed: _externalAuths.collectionId, _externalAuths.recordId, _externalAuths.provider (2067)",
  "errorMessage": "Failed to authenticate."
}
```

Please note that this bug is sporadic. I recommend attempting to log in multiple times and across different web browsers to replicate it 🙏🏽 .

**Reply by ganigeorgiev** [reply]
It is not a bug with PocketBase because it is working fine locally. As mentioned in https://github.com/pocketbase/pocketbase/discussions/966#discussioncomment-4088942, to rule out PocketBase issue, you can try replicating the static HTML example in your `pb_public` dir.

I have very little to none experience with Remix but if your code is executed on the server-side in a long running node process, then the global PocketBase SDK instance in https://github.com/hzhu/remix-pocketbase-oauth/blob/main/app/database.ts is incorrect. When using SSR (_in a long running node process_) each request must have their own SDK instance otherwise you are risking requests from different users to write over the same `pb.authStore`.

**Reply by ganigeorgiev** [reply]
I also forgot to mention that the above UNIQUE error indicates that you are trying to authenticate with more than 1 Google account for a single record, which is not supported, aka. you can have only 1 unique user-provider relation associated for a single user.

The mixup with the different Google accounts could happened due to the global SDK instance.
Keep in mind that if the user was already authenticated (aka. `pb.authStore.token` is not empty before making the OAuth2 call), we'll try to link the OAuth2 user to the user associated with the already authenticated token (this works only if the OAuth2 account wasn't previously linked to another PocketBase user).

**Reply by hzhu** [reply]
> When using SSR (in a long running node process) each request must have their own SDK instance otherwise you are risking requests from different users to write over the same `pb.authStore`.

I think that was the issue! I instantiated a new SDK instance on a per-request basis, and it works 🎉. Are there any drawbacks or other considerations I should be aware of when instantiating a new SDK instance for each request? Thank you @ganigeorgiev 🙇🏽 .

**Reply by ganigeorgiev** [reply]
> Are there any drawbacks or other considerations I should be aware of when instantiating a new SDK instance for each request? 

No, it is just not as convenient as having a global instance, but other than that it should be fine.
The JS SDK instance doesn't hold much data and has mostly functions/methods attached to it, so the memory consumption shouldn't be much different compared to creating a regular JS object.
